core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -811,31 +821,40 @@ class KafkaApis(val requestChannel: RequestChannel,
           .setRecords(data.records)
           .setPreferredReadReplica(data.preferredReadReplica.getOrElse(FetchResponse.INVALID_PREFERRED_REPLICA_ID))
         data.divergingEpoch.foreach(partitionData.setDivergingEpoch)
-        partitions.put(tp, partitionData)
+        addPartition(tp.topic, partitionData)
       }
-      erroneous.foreach { case (tp, data) => partitions.put(tp, data) }
-
-      var unconvertedFetchResponse: FetchResponse = null
+      erroneous.foreach { case (tp, data) => addPartition(tp.topic, data) }
 
-      def createResponse(throttleTimeMs: Int): FetchResponse = {
+      def createResponse(unconvertedFetchResponse: FetchResponse, throttleTimeMs: Int): FetchResponse = {
         // Down-convert messages for each partition if required
-        val convertedData = new util.LinkedHashMap[TopicPartition, FetchResponseData.PartitionData]
-        unconvertedFetchResponse.responseData.forEach { (tp, unconvertedPartitionData) =>
-          val error = Errors.forCode(unconvertedPartitionData.errorCode)
-          if (error != Errors.NONE)
-            debug(s"Fetch request with correlation id ${request.header.correlationId} from client $clientId " +
-              s"on partition $tp failed due to ${error.exceptionName}")
-          convertedData.put(tp, maybeConvertFetchedData(tp, unconvertedPartitionData))
+        val convertedResponse = new FetchResponseData()
+          .setErrorCode(unconvertedFetchResponse.error.code)
+          .setThrottleTimeMs(throttleTimeMs)
+          .setSessionId(unconvertedFetchResponse.sessionId)
+          .setResponses(new util.ArrayList[FetchResponseData.FetchableTopicResponse](unconvertedFetchResponse.data.responses.size))
+        unconvertedFetchResponse.data.responses.forEach { unconvertedTopicData =>
+          val convertedTopicResponse = new FetchResponseData.FetchableTopicResponse()
+            .setTopic(unconvertedTopicData.topic)
+            .setPartitions(new util.ArrayList[FetchResponseData.PartitionData](unconvertedTopicData.partitions.size))
+          convertedResponse.responses.add(convertedTopicResponse)
+          unconvertedTopicData.partitions.forEach { unconvertedPartitionData =>
+            val tp = new TopicPartition(unconvertedTopicData.topic, unconvertedPartitionData.partitionIndex)
```

Comment:
It's a pity that we have te recreate TopicPartition object here. We have to do the same in couple of other places... I am not sure about the impact though. What's your take on this?

I wonder if we could cache it in the response. We could extend the automated protocol to support "transient fields" (a field which is declared but not serialized) and store it there.

Reply:
yep, we made many copies when processing fetch request.

1. merge responsePartitionData with errors
1. down-convert fetch data 
1. regroup data

This patch want to fix 3. and others are traced by https://issues.apache.org/jira/browse/KAFKA-12387

At any rate, I will take a look at other cases if it does not bring a big patch.